gotwebd/gotwebd.8
.\"
.\" Copyright (c) 2020 Stefan Sperling
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt GOTWEBD 8
.Os
.Sh NAME
.Nm gotwebd
.Nd Game of Trees FastCGI server for web browsers
.Sh SYNOPSIS
.Nm
.Op Fl dnv
.Op Fl D Ar macro Ns = Ns Ar value
.Op Fl f Ar file
.Sh DESCRIPTION
.Nm
is a FastCGI server program which can display the contents of Git
repositories via a web browser.
The program has been designed to work out of the box with
the
.Xr httpd 8
web server.
.Pp
.Nm
provides the following options:
.Bl -tag -width tenletters
.It Fl D Ar macro Ns = Ns Ar value
Define
.Ar macro
to be set to
.Ar value .
Overrides the definition of
.Ar macro
in the configuration file.
.It Fl d
Do not daemonize.
Send log output to stderr.
.It Fl f Ar file
Set the path to the configuration file.
If not specified, the file
.Pa /etc/gotwebd.conf
will be used.
.It Fl n
Parse the configuration file, report errors if any, and exit.
.It Fl v
Verbose mode.
Verbosity increases if this option is used multiple times.
.El
.Pp
Enabling
.Nm
requires the following steps:
.Bl -enum
.It
The
.Xr httpd.conf 5
configuration file must be adjusted to run
.Nm
as a FastCGI helper program.
The
.Sx EXAMPLES
section below contains an appropriate configuration file sample.
.It
httpd(8) must be enabled and started:
.Bd -literal -offset indent
  # rcctl enable httpd
  # rcctl start httpd
.Ed
.It
Optionally, the run-time behaviour of
.Nm
can be configured via the
.Xr gotwebd.conf 5
configuration file.
.It
Git repositories must be created.
These repositories may reside anywhere in the filesystem and must
be readable, but should
.Em not
be writable, by the user
.Nm
runs as.
The default location for repositories published by
.Nm
is
.Pa /var/www/got/public .
.It
If the Git repositories served by
.Nm
do not receive changes from committers directly, they need to be kept
up-to-date with a mechanism such as
.Cm got fetch ,
.Xr git-fetch 1 ,
or
.Xr rsync 1 ,
scheduled by
.Xr cron 8 .
.El
.Sh FILES
.Bl -tag -width /var/www/got/public/ -compact
.It Pa /etc/gotwebd.conf
Default location of the
.Xr gotwebd.conf 5
configuration file.
.It Pa /var/www/got/public/
Default location for Git repositories served by
.Nm .
This location can be adjusted in the
.Xr gotwebd.conf 5
configuration file.
.It Pa /var/www/bin/gotwebd/
Directory containing statically linked
.Xr got 1
helper programs which are run by
.Nm
to read Git repositories.
.It Pa /var/www/htdocs/gotwebd/
Directory containing HTML, CSS, and image files used by
.Nm .
.It Pa /var/www/run/gotweb.sock
Default location for the
.Nm
listening socket.
.It Pa /tmp/
Directory for temporary files created by
.Nm .
.El
.Sh EXAMPLES
Example configuration for
.Xr httpd.conf 5 :
.Bd -literal -offset indent
types { include "/usr/share/misc/mime.types" }

server "example.com" {
	listen on * port 80
	root "/htdocs/gotwebd"
	location "/" {
		fastcgi socket "/run/gotweb.sock"
	}
}
.Ed
.Pp
Hosting multiple
.Nm gotwebd
instances on the same HTTP server under different path prefixes, with
the first reached via the default
.Ux Ns -domain socket, the second configured to listen on localhost
port 9000:
.Bd -literal -offset indent
server "example.com" {
	listen on * port 80

	location "/gotwebd-unix/" {
		fastcgi socket "/run/gotweb.sock"
	}
	location "/gotwebd-unix/*" {
		root "/htdocs/gotwebd"
		request strip 1
	}

	location "/gotwebd-tcp/" {
		fastcgi socket tcp localhost 9000
	}
	location "/gotwebd-tcp/*" {
		root "/htdocs/gotwebd"
		request strip 1
	}
}
.Ed
.Sh SEE ALSO
.Xr got 1 ,
.Xr git-repository 5 ,
.Xr gotwebd.conf 5 ,
.Xr httpd.conf 5 ,
.Xr httpd 8
.Sh AUTHORS
.An Omar Polo Aq Mt op@openbsd.org
.An Stefan Sperling Aq Mt stsp@openbsd.org
.An Tracey Emery Aq Mt tracey@traceyemery.net


CHANGES
* got 0.110; 2025-02-20
- fix an endless loop in got-read-pack (regression from 0.109)
- change gotwebd diff algorithm from Myers to Patience diff
- gotd regress depends on p5-Net-Daemon now due to an added test case

* got 0.109; 2025-02-14
- fix gotd failing to protect references when the client sends an empty pack
- during pack generation, fix exclusion of commits via an ancestor commit
- fix a bogus "received unexpected privsep message" error from gotsh
- fix diffstat path order bug in field width computation
- gotwebd: preserve 'folder=' parameter when following More links

* got 0.108; 2025-01-22
- add ssh -i identity-file support to commands which use the network
- make 'got import' output independent of readdir(3) entry order
- avoid full file content comparisons in 'got status' for speed
- tog: fix NULL deref when log view T keymap is used on worktree entry
- tog: fix a deadlock (hang) in the log view implementation
- tog: plug a memory leak
- tog: do not exit if a tag pointing at a non-commit is selected in ref view
- tog: do not mark an incorrect base commit in nested log views
- tog: fix NULL deref when scrolling small tree views down
- tog: avoid showing a negative log view entry index
- tog: do not apply a pointless count modifier to the H, &, p keymaps
- tog: do not make users wait for the worktree diff to quit out of tog
- gotwebd: make parent process drop root privileges
- gotwebd: drop read access to /var/www from parent process
- gotwebd: rename "socket" processes to "server"
- gotadmin cleanup: pack the repository before removing objects
- gotadmin cleanup: do not delete directly referenced trees and blobs
- gotadmin cleanup: do not delete objects reachable via nested tags
- regress: skip test memleak_send_basic in sha256 mode; expected to fail
- regress: make seq(1) invocations portable to fix test failures on linux
- regress/gotwebd: implement paginated commits test

* got 0.107; 2024-12-28
- gotwebd.css styling tweaks
- hide ssh debug output during fetch/send -v, keep showing it at -vv and -vvv
- discern mixed-commit worktree diffs with commit ID headers
- gotwebd: avoid printf("%s", NULL) when path parameter is not in query
- implement a regression test harness for gotwebd
- fix free() called with bogus pointer in 'got fetch'; regression from 0.106
- ensure config privsep children get collected upon error to prevent zombies
- fix some fprintf(3) failure checks
- gotwebd: replace strftime(3) with asctime_r(3) for the sake of consistency
- tweak gotwebd log message levels, and log requests in verbose (-v) mode
- prevent out-of-bounds read during gotwebd fcgi record debugging
- implement tog work tree diff support via log view and CLI
- improve error reporting when 'got patch' encounters malformed patches
- improve got_opentemp_named_fd error reporting by showing the path template
- add ssh -J jumphost support to got and cvg commands which use the network
- add regression tests checking for memory leaks with Otto malloc and ktrace
- got tag: change -s signer to -S signer
- got tag: provide one-line output mode via new -s option
- tog: use wtimeout(3) instead of nodelay(3) to honour our display refresh rate 
- switch got_pathlist data store from TAILQ to RB-tree
- plug many memory leaks, some of which affected gotwebd in particular

* got 0.106; 2024-11-21
- prevent gotd from exiting with pending notifications if client disconnects
- convert got to the new imsg API
- gotwebd: improve performance of repository age calculations
- gotwebd: ensure child processes inherit non-default config

* got 0.105; 2024-11-14
- fix bogus "branch on server has different ancestry" errors from 'got send'
- do not try to merge binary files during (un)stage -p and revert -p
- fix gotd notifications about changes involving empty files
- minor gotwebd.css styling changes
- plug several memory and file-descriptor leaks

* got 0.104; 2024-10-22
  see git repository history for per-change authorship information
- gotd.conf: document the macro syntax
- tog: prevent a segfault upon unexpected object type in ref list view
- fix pack file creation in the presence of tagged tag objects
- plugged some memory leaks
- fix a crash when unstaging a file which has been removed from disk
- gotwebd: fix out of bounds access while handling the configuration

* got 0.103; 2024-09-24
  see git repository history for per-change authorship information
- fix bug causing performance to degrade as more and more pack files appear
- tog: add diff view 'p' keymap to write the diff to file
- tog: display diffstat in diff view when diffing blobs or trees directly
- gotwebd: show commit id prefix on briefs page
- add support for HMAC digests to gotd HTTP notifications
- move authentication credentials from gotd.conf(5) to gotd-secrets.conf(5)
- fix spurious tog regression test failures on slower machines
- restore abort() calls in lib/hash.c to quiet potential compiler warnings
- gotwebd: unbreak listing of tags on the summary page (regression from 0.102)
- gotwebd: minor tweaks to the HTML for ease of styling

* got 0.102; 2024-08-14
  see git repository history for per-change authorship information
- support for sha256 repositories; the network protocol requires git(1) for now
- gotwebd: add support for the "owner" file
- gotwebd: fix the README link in the summary view
- fix handling of .gitignore files containing empty lines
- fix handling of files without trailing newline in histedit, rebase and merge
- gotd: allow numeric UIDs in permit/deny rules, as intended
- gotd: support numeric UIDs in the `user' directive
- fix comment handling and explain quoting in the *.conf.5 man pages
- tog: add ability to mark arbitrary commits to diff them
- print file index and work tree version in got info

* got 0.101; 2024-07-11
  see git repository history for per-change authorship information
- improve gotwebd.8 one-line description
- adjust some SIZE_MAX bounds checks

gotwebd/gotwebd.c
/*
 * Copyright (c) 2016, 2019, 2020-2021 Tracey Emery <tracey@traceyemery.net>
 * Copyright (c) 2015 Reyk Floeter <reyk@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "got_compat.h"

#include <sys/param.h>
#include <sys/queue.h>
#include <sys/socket.h>
#include <sys/wait.h>

#include <net/if.h>
#include <netinet/in.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <err.h>
#include <errno.h>
#include <event.h>
#include <fcntl.h>
#include <pwd.h>
#include <signal.h>
#include <syslog.h>
#include <unistd.h>
#include <ctype.h>

#include "got_opentemp.h"
#include "got_reference.h"

#include "gotwebd.h"
#include "log.h"

__dead void usage(void);

int	 main(int, char **);
int	 gotwebd_configure(struct gotwebd *, uid_t, gid_t);
void	 gotwebd_configure_done(struct gotwebd *);
void	 gotwebd_sighdlr(int sig, short event, void *arg);
void	 gotwebd_shutdown(void);
void	 gotwebd_dispatch_server(int, short, void *);
void	 gotwebd_dispatch_gotweb(int, short, void *);

struct gotwebd	*gotwebd_env;

void
imsg_event_add(struct imsgev *iev)
{
	if (iev->handler == NULL) {
		imsgbuf_flush(&iev->ibuf);
		return;
	}

	iev->events = EV_READ;
	if (imsgbuf_queuelen(&iev->ibuf))
		iev->events |= EV_WRITE;

	event_del(&iev->ev);
	event_set(&iev->ev, iev->ibuf.fd, iev->events, iev->handler, iev->data);
	event_add(&iev->ev, NULL);
}

int
imsg_compose_event(struct imsgev *iev, uint16_t type, uint32_t peerid,
    pid_t pid, int fd, const void *data, size_t datalen)
{
	int ret;

	ret = imsg_compose(&iev->ibuf, type, peerid, pid, fd, data, datalen);
	if (ret == -1)
		return (ret);
	imsg_event_add(iev);
	return (ret);
}

static int
send_imsg(struct imsgev *iev, uint32_t type, int fd, const void *data,
    uint16_t len)
{
	int	 ret, d = -1;

	if (fd != -1 && (d = dup(fd)) == -1)
		goto err;

	ret = imsg_compose_event(iev, type, 0, -1, d, data, len);
	if (ret == -1)
		goto err;

	if (d != -1) {
		d = -1;
		/* Flush imsg to prevent fd exhaustion. 'd' will be closed. */
		if (imsgbuf_flush(&iev->ibuf) == -1)
			goto err;
		imsg_event_add(iev);
	}

	return 0;
err:
	if (d != -1)
		close(d);
	return -1;
}

int
main_compose_sockets(struct gotwebd *env, uint32_t type, int fd,
    const void *data, uint16_t len)
{
	size_t i;
	int ret = 0;

	for (i = 0; i < env->nserver; ++i) {
		ret = send_imsg(&env->iev_server[i], type, fd, data, len);
		if (ret)
			break;
	}

	return ret;
}

int
main_compose_gotweb(struct gotwebd *env, uint32_t type, int fd,
    const void *data, uint16_t len)
{
	size_t i;
	int ret = 0;

	for (i = 0; i < env->nserver; ++i) {
		ret = send_imsg(&env->iev_gotweb[i], type, fd, data, len);
		if (ret)
			break;
	}

	return ret;
}

int
sockets_compose_main(struct gotwebd *env, uint32_t type, const void *d,
    uint16_t len)
{
	return (imsg_compose_event(env->iev_parent, type, 0, -1, -1, d, len));
}

void
gotwebd_dispatch_server(int fd, short event, void *arg)
{
	struct imsgev		*iev = arg;
	struct imsgbuf		*ibuf;
	struct imsg		 imsg;
	struct gotwebd		*env = gotwebd_env;
	ssize_t			 n;
	int			 shut = 0;

	ibuf = &iev->ibuf;

	if (event & EV_READ) {
		if ((n = imsgbuf_read(ibuf)) == -1)
			fatal("imsgbuf_read error");
		if (n == 0)	/* Connection closed */
			shut = 1;
	}
	if (event & EV_WRITE) {
		if (imsgbuf_write(ibuf) == -1)
			fatal("imsgbuf_write");
	}

	for (;;) {
		if ((n = imsg_get(ibuf, &imsg)) == -1)
			fatal("imsg_get");
		if (n == 0)	/* No more messages. */
			break;

		switch (imsg.hdr.type) {
		case GOTWEBD_IMSG_CFG_DONE:
			gotwebd_configure_done(env);
			break;
		default:
			fatalx("%s: unknown imsg type %d", __func__,
			    imsg.hdr.type);
		}

		imsg_free(&imsg);
	}

	if (!shut)
		imsg_event_add(iev);
	else {
		/* This pipe is dead.  Remove its event handler */
		event_del(&iev->ev);
		event_loopexit(NULL);
	}
}

void
gotwebd_dispatch_gotweb(int fd, short event, void *arg)
{
	struct imsgev		*iev = arg;
	struct imsgbuf		*ibuf;
	struct imsg		 imsg;
	struct gotwebd		*env = gotwebd_env;
	ssize_t			 n;
	int			 shut = 0;

	ibuf = &iev->ibuf;

	if (event & EV_READ) {
		if ((n = imsgbuf_read(ibuf)) == -1)
			fatal("imsgbuf_read error");
		if (n == 0)	/* Connection closed */
			shut = 1;
	}
	if (event & EV_WRITE) {
		if (imsgbuf_write(ibuf) == -1)
			fatal("imsgbuf_write");
	}

	for (;;) {
		if ((n = imsg_get(ibuf, &imsg)) == -1)
			fatal("imsg_get");
		if (n == 0)	/* No more messages. */
			break;

		switch (imsg.hdr.type) {
		case GOTWEBD_IMSG_CFG_DONE:
			gotwebd_configure_done(env);
			break;
		default:
			fatalx("%s: unknown imsg type %d", __func__,
			    imsg.hdr.type);
		}

		imsg_free(&imsg);
	}

	if (!shut)
		imsg_event_add(iev);
	else {
		/* This pipe is dead.  Remove its event handler */
		event_del(&iev->ev);
		event_loopexit(NULL);
	}
}

void
gotwebd_sighdlr(int sig, short event, void *arg)
{
	/* struct privsep	*ps = arg; */

	switch (sig) {
	case SIGHUP:
		log_info("%s: ignoring SIGHUP", __func__);
		break;
	case SIGPIPE:
		log_info("%s: ignoring SIGPIPE", __func__);
		break;
	case SIGUSR1:
		log_info("%s: ignoring SIGUSR1", __func__);
		break;
	case SIGTERM:
	case SIGINT:
		gotwebd_shutdown();
		break;
	default:
		fatalx("unexpected signal");
	}
}

static void
spawn_process(struct gotwebd *env, const char *argv0, struct imsgev *iev,
    enum gotwebd_proc_type proc_type, const char *username,
    void (*handler)(int, short, void *))
{
	const char	*argv[9];
	int		 argc = 0;
	int		 p[2];
	pid_t		 pid;

	if (socketpair(AF_UNIX, SOCK_STREAM, PF_UNSPEC, p) == -1)
		fatal("socketpair");

	switch (pid = fork()) {
	case -1:
		fatal("fork");
	case 0:		/* child */
		break;
	default:	/* parent */
		close(p[0]);
		if (imsgbuf_init(&iev->ibuf, p[1]) == -1)
			fatal("imsgbuf_init");
		imsgbuf_allow_fdpass(&iev->ibuf);
		iev->handler = handler;
		iev->data = iev;
		event_set(&iev->ev, p[1], EV_READ, handler, iev);
		event_add(&iev->ev, NULL);
		return;
	}

	close(p[1]);

	argv[argc++] = argv0;
	if (proc_type == GOTWEBD_PROC_SERVER) {
		argv[argc++] = "-S";
		argv[argc++] = username;
	} else if (proc_type == GOTWEBD_PROC_GOTWEB) {
		argv[argc++] = "-G";
		argv[argc++] = username;
	}
	if (strcmp(env->gotwebd_conffile, GOTWEBD_CONF) != 0) {
		argv[argc++] = "-f";
		argv[argc++] = env->gotwebd_conffile;
	}
	if (env->gotwebd_debug)
		argv[argc++] = "-d";
	if (env->gotwebd_verbose > 0)
		argv[argc++] = "-v";
	if (env->gotwebd_verbose > 1)
		argv[argc++] = "-v";
	argv[argc] = NULL;

	if (p[0] != GOTWEBD_SOCK_FILENO) {
		if (dup2(p[0], GOTWEBD_SOCK_FILENO) == -1)
			fatal("dup2");
	} else if (fcntl(p[0], F_SETFD, 0) == -1)
		fatal("fcntl");

	/* obnoxious cast */
	execvp(argv0, (char * const *)argv);
	fatal("execvp %s", argv0);
}

__dead void
usage(void)
{
	fprintf(stderr, "usage: %s [-dnv] [-D macro=value] [-f file]\n",
	    getprogname());
	exit(1);
}

int
main(int argc, char **argv)
{
	struct event		 sigint, sigterm, sighup, sigpipe, sigusr1;
	struct event_base	*evb;
	struct gotwebd		*env;
	struct passwd		*pw;
	int			 ch, i;
	int			 no_action = 0;
	int			 proc_type = GOTWEBD_PROC_PARENT;
	const char		*conffile = GOTWEBD_CONF;
	const char		*gotwebd_username = GOTWEBD_DEFAULT_USER;
	const char		*www_username = GOTWEBD_WWW_USER;
	gid_t			 www_gid;
	const char		*argv0;

	if ((argv0 = argv[0]) == NULL)
		argv0 = "gotwebd";

	/* log to stderr until daemonized */
	log_init(1, LOG_DAEMON);

	env = calloc(1, sizeof(*env));
	if (env == NULL)
		fatal("%s: calloc", __func__);
	config_init(env);

	while ((ch = getopt(argc, argv, "D:dG:f:nS:vW:")) != -1) {
		switch (ch) {
		case 'D':
			if (cmdline_symset(optarg) < 0)
				log_warnx("could not parse macro definition %s",
				    optarg);
			break;
		case 'd':
			env->gotwebd_debug = 1;
			break;
		case 'G':
			proc_type = GOTWEBD_PROC_GOTWEB;
			gotwebd_username = optarg;
			break;
		case 'f':
			conffile = optarg;
			break;
		case 'n':
			no_action = 1;
			break;
		case 'S':
			proc_type = GOTWEBD_PROC_SERVER;
			gotwebd_username = optarg;
			break;
		case 'v':
			if (env->gotwebd_verbose < 3)
				env->gotwebd_verbose++;
			break;
		default:
			usage();
		}
	}

	argc -= optind;
	if (argc > 0)
		usage();

	gotwebd_env = env;
	env->gotwebd_conffile = conffile;

	if (proc_type == GOTWEBD_PROC_PARENT) {
		if (parse_config(env->gotwebd_conffile, env) == -1)
			exit(1);

		if (no_action) {
			fprintf(stderr, "configuration OK\n");
			exit(0);
		}

		if (env->user)
			gotwebd_username = env->user;
		if (env->www_user)
			www_username = env->www_user;
	}

	pw = getpwnam(www_username);
	if (pw == NULL)
		fatalx("unknown user %s", www_username);
	www_gid = pw->pw_gid;

	pw = getpwnam(gotwebd_username);
	if (pw == NULL)
		fatalx("unknown user %s", gotwebd_username);

	/* check for root privileges */
	if (geteuid())
		fatalx("need root privileges");

	log_init(env->gotwebd_debug, LOG_DAEMON);
	log_setverbose(env->gotwebd_verbose);

	switch (proc_type) {
	case GOTWEBD_PROC_SERVER:
		setproctitle("server");
		log_procinit("server");

		if (chroot(env->httpd_chroot) == -1)
			fatal("chroot %s", env->httpd_chroot);
		if (chdir("/") == -1)
			fatal("chdir /");

		if (setgroups(1, &pw->pw_gid) == -1 ||
		    setresgid(pw->pw_gid, pw->pw_gid, pw->pw_gid) == -1 ||
		    setresuid(pw->pw_uid, pw->pw_uid, pw->pw_uid) == -1)
			fatal("failed to drop privileges");

		sockets(env, GOTWEBD_SOCK_FILENO);
		return 1;
	case GOTWEBD_PROC_GOTWEB:
		setproctitle("gotweb");
		log_procinit("gotweb");

		if (setgroups(1, &pw->pw_gid) == -1 ||
		    setresgid(pw->pw_gid, pw->pw_gid, pw->pw_gid) == -1 ||
		    setresuid(pw->pw_uid, pw->pw_uid, pw->pw_uid) == -1)
			fatal("failed to drop privileges");

		gotweb(env, GOTWEBD_SOCK_FILENO);
		return 1;
	default:
		break;
	}

	if (!env->gotwebd_debug && daemon(1, 0) == -1)
		fatal("daemon");

	evb = event_init();

	env->nserver = env->prefork_gotwebd;
	env->iev_server = calloc(env->nserver, sizeof(*env->iev_server));
	if (env->iev_server == NULL)
		fatal("calloc");
	env->iev_gotweb = calloc(env->nserver, sizeof(*env->iev_gotweb));
	if (env->iev_gotweb == NULL)
		fatal("calloc");

	for (i = 0; i < env->nserver; ++i) {
		spawn_process(env, argv0, &env->iev_server[i],
		    GOTWEBD_PROC_SERVER, gotwebd_username,
		    gotwebd_dispatch_server);
		spawn_process(env, argv0, &env->iev_gotweb[i],
		    GOTWEBD_PROC_GOTWEB, gotwebd_username,
		    gotwebd_dispatch_gotweb);
	}
	if (chdir("/") == -1)
		fatal("chdir /");

	log_procinit("gotwebd");

	log_info("%s startup", getprogname());

	signal_set(&sigint, SIGINT, gotwebd_sighdlr, env);
	signal_set(&sigterm, SIGTERM, gotwebd_sighdlr, env);
	signal_set(&sighup, SIGHUP, gotwebd_sighdlr, env);
	signal_set(&sigpipe, SIGPIPE, gotwebd_sighdlr, env);
	signal_set(&sigusr1, SIGUSR1, gotwebd_sighdlr, env);

	signal_add(&sigint, NULL);
	signal_add(&sigterm, NULL);
	signal_add(&sighup, NULL);
	signal_add(&sigpipe, NULL);
	signal_add(&sigusr1, NULL);

	if (gotwebd_configure(env, pw->pw_uid, www_gid) == -1)
		fatalx("configuration failed");

	if (setgroups(1, &pw->pw_gid) == -1 ||
	    setresgid(pw->pw_gid, pw->pw_gid, pw->pw_gid) == -1 ||
	    setresuid(pw->pw_uid, pw->pw_uid, pw->pw_uid) == -1)
		fatal("failed to drop privileges");

#ifdef PROFILE
	if (unveil("gmon.out", "rwc") != 0)
		err(1, "gmon.out");
#endif

	if (unveil(GOTWEBD_CONF, "r") == -1)
		err(1, "unveil");

	if (unveil(NULL, NULL) != 0)
		err(1, "unveil");

#ifndef PROFILE
	if (pledge("stdio", NULL) == -1)
		err(1, "pledge");
#endif

	event_dispatch();
	event_base_free(evb);

	log_debug("%s gotwebd exiting", getprogname());

	return (0);
}

static void
connect_children(struct gotwebd *env)
{
	struct imsgev *iev1, *iev2;
	int pipe[2];
	int i;

	for (i = 0; i < env->nserver; i++) {
		iev1 = &env->iev_server[i];
		iev2 = &env->iev_gotweb[i];
	
		if (socketpair(AF_UNIX, SOCK_STREAM, PF_UNSPEC, pipe) == -1)
			fatal("socketpair");

		if (send_imsg(iev1, GOTWEBD_IMSG_CTL_PIPE, pipe[0], NULL, 0))
			fatal("send_imsg");

		if (send_imsg(iev2, GOTWEBD_IMSG_CTL_PIPE, pipe[1], NULL, 0))
			fatal("send_imsg");

		close(pipe[0]);
		close(pipe[1]);
	}
}

int
gotwebd_configure(struct gotwebd *env, uid_t uid, gid_t gid)
{
	struct server *srv;
	struct socket *sock;

	/* gotweb need to reload its config. */
	env->servers_pending = env->prefork_gotwebd;
	env->gotweb_pending = env->prefork_gotwebd;

	/* send our gotweb servers */
	TAILQ_FOREACH(srv, &env->servers, entry) {
		if (main_compose_sockets(env, GOTWEBD_IMSG_CFG_SRV,
		    -1, srv, sizeof(*srv)) == -1)
			fatal("main_compose_sockets GOTWEBD_IMSG_CFG_SRV");
		if (main_compose_gotweb(env, GOTWEBD_IMSG_CFG_SRV,
		    -1, srv, sizeof(*srv)) == -1)
			fatal("main_compose_gotweb GOTWEBD_IMSG_CFG_SRV");
	}

	/* send our sockets */
	TAILQ_FOREACH(sock, &env->sockets, entry) {
		if (config_setsock(env, sock, uid, gid) == -1)
			fatalx("%s: send socket error", __func__);
	}

	/* send the temp files */
	if (config_setfd(env) == -1)
		fatalx("%s: send priv_fd error", __func__);

	/* Connect servers and gotwebs. */
	connect_children(env);

	if (main_compose_sockets(env, GOTWEBD_IMSG_CFG_DONE, -1, NULL, 0) == -1)
		fatal("main_compose_sockets GOTWEBD_IMSG_CFG_DONE");

	return (0);
}

void
gotwebd_configure_done(struct gotwebd *env)
{
	if (env->servers_pending > 0) {
		env->servers_pending--;
		if (env->servers_pending == 0 &&
		    main_compose_sockets(env, GOTWEBD_IMSG_CTL_START,
		        -1, NULL, 0) == -1)
			fatal("main_compose_sockets GOTWEBD_IMSG_CTL_START");
	}

	if (env->gotweb_pending > 0) {
		env->gotweb_pending--;
		if (env->gotweb_pending == 0 &&
		    main_compose_gotweb(env, GOTWEBD_IMSG_CTL_START,
		        -1, NULL, 0) == -1)
			fatal("main_compose_sockets GOTWEBD_IMSG_CTL_START");
	}
}

void
gotwebd_shutdown(void)
{
	struct gotwebd	*env = gotwebd_env;
	pid_t		 pid;
	int		 i, status;

	for (i = 0; i < env->nserver; ++i) {
		event_del(&env->iev_server[i].ev);
		imsgbuf_clear(&env->iev_server[i].ibuf);
		close(env->iev_server[i].ibuf.fd);
		env->iev_server[i].ibuf.fd = -1;
	}
	free(env->iev_server);

	for (i = 0; i < env->nserver; ++i) {
		event_del(&env->iev_gotweb[i].ev);
		imsgbuf_clear(&env->iev_gotweb[i].ibuf);
		close(env->iev_gotweb[i].ibuf.fd);
		env->iev_gotweb[i].ibuf.fd = -1;
	}
	free(env->iev_gotweb);

	do {
		pid = waitpid(WAIT_ANY, &status, 0);
		if (pid <= 0)
			continue;

		if (WIFSIGNALED(status))
			log_warnx("lost child: pid %u terminated; signal %d",
			    pid, WTERMSIG(status));
		else if (WIFEXITED(status) && WEXITSTATUS(status) != 0)
			log_warnx("lost child: pid %u exited abnormally",
			    pid);
	} while (pid != -1 || (pid == -1 && errno == EINTR));

	while (!TAILQ_EMPTY(&gotwebd_env->addresses)) {
		struct address *h;

		h = TAILQ_FIRST(&gotwebd_env->addresses);
		TAILQ_REMOVE(&gotwebd_env->addresses, h, entry);
		free(h);
	}
	while (!TAILQ_EMPTY(&gotwebd_env->servers)) {
		struct server *srv;

		srv = TAILQ_FIRST(&gotwebd_env->servers);
		TAILQ_REMOVE(&gotwebd_env->servers, srv, entry);
		free(srv);
	}
	sockets_purge(gotwebd_env);
	free(gotwebd_env);

	log_warnx("gotwebd terminating");
	exit(0);
}
